### PR TITLE
KEYCLOAK-16814 Deny access to the metrics endpoint

### PR DESCRIPTION
**Pull request by slaskawi**
## JIRA ID

https://issues.redhat.com/browse/KEYCLOAK-16814

## Additional Information

This Pull Request redirects all calls to the metrics endpoint to the master realm info.

## Checklist:
- [x] Automated Tests
- [x] [Documentation](https://github.com/keycloak/keycloak-documentation) changes if necessary
- [x] (If required) Discussed on [Keycloak DEV](https://groups.google.com/forum/#!forum/keycloak-dev)

Tagging for visibility @miquelsi @pb82 @davidffrench

### CODE REVIEW

**Comment by slaskawi** [comment]
@miquelsi @vmuzikar @abstractj Can I ask for a review?

File: pkg/model/keycloak_metrics_route.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+			Host: keycloakMainRouteCopy.Spec.Host,
+			Path: KeycloakMetricsRoutePath,
+			Port: keycloakMainRouteCopy.Spec.Port,
+			TLS:  keycloakMainRouteCopy.Spec.TLS,
+			To:   keycloakMainRouteCopy.Spec.To,
+		},
+	}
+}
+
+func KeycloakMetricsRouteReconciled(cr *kc.Keycloak, currentState *v1.Route, keycloakMainRoute *v1.Route) *v1.Route {
```
Comment (vmuzikar) [inline pkg/model/keycloak_metrics_route.go:34]: The `cr` argument is unused here but I guess it's to keep it consistent with other reconciler models.

**Review by vmuzikar — COMMENTED** [review]
@slaskawi Sorry for a late review.

Changes look good but I didn't try them. I presume you did.

I assume the `/metrics` endpoint is available only through the master realm, correct?

One thing I'm wondering about is some tests. It'd make sense to test a correct rewrite is applied when using OCP and K8s (which is addressed in #292) respectively.

File: pkg/model/keycloak_metrics_route.go (hunk 1 above)
Comment (slaskawi) [reply pkg/model/keycloak_metrics_route.go:34]: Yes, exactly. I like keeping them consistent. Oftentimes you copy-paste stuff around in the model package.

**Comment by slaskawi** [comment]
> Changes look good but I didn't try them. I presume you did.

@vmuzikar Yes. The OpenShift Router team helped me a lot with this solution (it's actually their idea!).

> I assume the /metrics endpoint is available only through the master realm, correct?

@vmuzikar Yes.

> One thing I'm wondering about is some tests. It'd make sense to test a correct rewrite is applied when using OCP and K8s (which is addressed in #292) respectively.

@vmuzikar Since I unified the approach, the test from https://github.com/keycloak/keycloak-operator/pull/312 should work on OpenShift as well.

**Comment by slaskawi** [comment]
@abstractj May I ask for a review?

**Review by vmuzikar — APPROVED** [review]

**Review by pedroigor — APPROVED** [review]

**Pull request merged**